File: apps/windows/TeslaSync.App/dashboard-widgets/BatteryCellsWidget/BatteryCellsWidget.Model.cs
using System.Globalization;
using System.Text.Json;
using TeslaSync.App.Core;
using TeslaSync.App.Core.Data.State;
using TeslaSync.App.Core.Notifications;
using TeslaSync.App.Core.Units;

namespace TeslaSync.App.DashboardWidgets;

/// <summary>
/// The lifecycle state a <see cref="BatteryCellsViewModel"/> can be in — the native union of the
/// loading / loaded / empty / error / stale / offline branches the web <c>BatteryCellsWidget</c>
/// renders through <c>WidgetShell</c>
/// (web/src/features/dashboard/widgets/BatteryCellsWidget.tsx). Every branch maps onto a visible
/// surface; none is ever hidden. <see cref="Empty"/> mirrors the web outer <c>{data ? … : &lt;EmptyState&gt;}</c>
/// gate (no resolved vehicle / no response body) — distinct from the inner empty grid (no cells),
/// which is rendered as part of <see cref="Loaded"/> with the "No cell data" message.
/// </summary>
public enum BatteryCellsState
{
    /// <summary>Initial fetch with no cached snapshot — render the skeleton chrome.</summary>
    Loading,

    /// <summary>A fresh snapshot (or non-stale cache) with a cell summary to show.</summary>
    Loaded,

    /// <summary>No vehicle resolved or no response body — render the "No battery cell data" empty state.</summary>
    Empty,

    /// <summary>The request failed and no cached snapshot exists — render the retry affordance.</summary>
    Error,

    /// <summary>A cached snapshot older than the freshness window — render content plus a stale chip.</summary>
    Stale,

    /// <summary>The network failed but a cached snapshot remains — render content plus an offline chip.</summary>
    Offline,
}

/// <summary>
/// The per-cell severity the voltage heatmap colours a tile with — the native union of the web
/// <c>StatusCell['status']</c> values the shared <c>WidgetStatusGrid</c> renders. Derived from how far
/// a cell's voltage deviates from the pack average (web <c>cellStatus</c>): ≤5&#160;mV → <see cref="Ok"/>,
/// ≤15&#160;mV → <see cref="Warning"/>, &gt;15&#160;mV → <see cref="Error"/>, missing voltage →
/// <see cref="Unknown"/>.
/// </summary>
public enum BatteryCellSeverity
{
    /// <summary>Within 5&#160;mV of the average (web <c>'ok'</c>).</summary>
    Ok,

    /// <summary>5–15&#160;mV from the average (web <c>'warning'</c>).</summary>
    Warning,

    /// <summary>More than 15&#160;mV from the average (web <c>'error'</c>).</summary>
    Error,

    /// <summary>No voltage reported for the cell (web <c>'unknown'</c>).</summary>
    Unknown,
}

/// <summary>
/// One battery brick reading from <c>GET /vehicles/{vehicleID}/battery/cells</c> — the native mirror of
/// the web <c>BatteryCell</c> type (web/src/types/energy.ts): identity (<c>cell_id</c>, <c>module</c>) plus
/// the measured <c>voltage</c> (volts, SI) and <c>temperature</c> (°C, SI). Parsing is null-tolerant so a
/// partial body never throws; <see cref="Voltage"/> stays null when the brick reported no voltage so the
/// heatmap can render the web <c>'unknown'</c> status rather than a fabricated value.
/// </summary>
public sealed record BatteryCell(int CellId, int Module, double? Voltage, double? Temperature)
{
    /// <summary>Project one cell object into a tolerant reading (binds to the web <c>BatteryCell</c> contract).</summary>
    public static BatteryCell FromJson(JsonElement element)
    {
        if (element.ValueKind != JsonValueKind.Object)
        {
            return new BatteryCell(0, 0, null, null);
        }

        return new BatteryCell(
            CellId: (int)Math.Round(JsonNum.Double(element, "cell_id") ?? 0),
            Module: (int)Math.Round(JsonNum.Double(element, "module") ?? 0),
            Voltage: JsonNum.Double(element, "voltage"),
            Temperature: JsonNum.Double(element, "temperature"));
    }
}

/// <summary>
/// The battery cell rollup from <c>GET /vehicles/{vehicleID}/battery/cells</c> (web <c>useBatteryCells</c>,
/// shape <c>BatteryCellSummary</c> in web/src/types/energy.ts). Field names mirror the Go API's snake_case
/// JSON tags. All voltages are volts and temperatures are °C (SI); they are formatted for display only at
/// projection time and never unit-converted (the web widget shows raw volts / °, matching this). Parsing is
/// null-tolerant so a partial body never throws.
/// </summary>
public sealed record BatteryCellSummary(
    double AvgVoltage,
    double MinVoltage,
    double MaxVoltage,
    double VoltageSpread,
    double AvgTemperature,
    double MinTemperature,
    double MaxTemperature,
    double TempSpread,
    int TotalCells,
    IReadOnlyList<BatteryCell> Cells)
{
    /// <summary>An all-zero summary with no cells — the parse fallback for an absent/non-object body.</summary>
    public static BatteryCellSummary Empty { get; } = new(
        0, 0, 0, 0, 0, 0, 0, 0, 0, Array.Empty<BatteryCell>());

    /// <summary>Project a <c>GET /vehicles/{vehicleID}/battery/cells</c> JSON object into a tolerant summary.</summary>
    public static BatteryCellSummary FromJson(JsonElement element)
    {
        if (element.ValueKind != JsonValueKind.Object)
        {
            return Empty;
        }

        return new BatteryCellSummary(
            AvgVoltage: JsonNum.Double(element, "avg_voltage") ?? 0,
            MinVoltage: JsonNum.Double(element, "min_voltage") ?? 0,
            MaxVoltage: JsonNum.Double(element, "max_voltage") ?? 0,
            VoltageSpread: JsonNum.Double(element, "voltage_spread") ?? 0,
            AvgTemperature: JsonNum.Double(element, "avg_temperature") ?? 0,
            MinTemperature: JsonNum.Double(element, "min_temperature") ?? 0,
            MaxTemperature: JsonNum.Double(element, "max_temperature") ?? 0,
            TempSpread: JsonNum.Double(element, "temp_spread") ?? 0,
            TotalCells: (int)Math.Round(JsonNum.Double(element, "total_cells") ?? 0),
            Cells: ReadCells(element));
    }

    private static IReadOnlyList<BatteryCell> ReadCells(JsonElement element)
    {
        if (!element.TryGetProperty("cells", out var cells) || cells.ValueKind != JsonValueKind.Array)
        {
            return Array.Empty<BatteryCell>();
        }

        var list = new List<BatteryCell>(cells.GetArrayLength());
        foreach (var cell in cells.EnumerateArray())
        {
            list.Add(BatteryCell.FromJson(cell));
        }

        return list;
    }
}

/// <summary>
/// The widget's grid footprint (columns × rows). Mirrors the web <c>WidgetProps.size</c> plus the
/// <c>isCompact</c> / <c>isWide</c> / heatmap-column logic in
/// web/src/features/dashboard/widgets/BatteryCellsWidget.tsx.
/// </summary>
public readonly record struct BatteryCellsSize(int Cols, int Rows)
{
    /// <summary>The registry default footprint (2×4).</summary>
    public static BatteryCellsSize Default => new(2, 4);

    /// <summary>True at a single column (web <c>isCompact = size.cols &lt;= 1</c>): hide the title, tighten the grid.</summary>
    public bool IsCompact => Cols <= 1;

    /// <summary>True at three or more columns (web <c>isWide = size.cols &gt;= 3</c>): per-cell temperature + a temp row.</summary>
    public bool IsWide => Cols >= 3;

    /// <summary>Heatmap column count (web <c>cols={isWide ? 4 : isCompact ? 2 : 3}</c>).</summary>
    public int GridColumns => IsWide ? 4 : IsCompact ? 2 : 3;
}

/// <summary>
/// One projected, display-ready heatmap tile consumed by the WinUI view — the native analogue of a web
/// <c>StatusCell</c>. Holds the localized label, the already-formatted value, the derived severity (which
/// the view maps to a themed status brush) and a Narrator automation name. Pure data — no WinUI types.
/// </summary>
public sealed record BatteryCellStatusItem(
    string Id,
    string Label,
    BatteryCellSeverity Severity,
    string Value,
    string AutomationName);

/// <summary>
/// One projected stat tile (min/max/avg/spread or per-module temperature) — the native analogue of a web
/// <c>StatCard</c>. Holds the localized label and the already-formatted value. Pure data.
/// </summary>
public sealed record BatteryCellsStat(string Label, string Value);

/// <summary>
/// The fully projected, render-ready view of the battery cell summary for one footprint — the native
/// analogue of everything the web component computes via <c>useMemo</c> before returning JSX. Holds the
/// heatmap tiles, the four voltage stats, and (when wide) the three temperature stats, plus the footprint
/// flags. Pure data so the projection is unit-tested without a UI host.
/// </summary>
public sealed record BatteryCellsDisplay(
    bool IsCompact,
    bool IsWide,
    int GridColumns,
    IReadOnlyList<BatteryCellStatusItem> Cells,
    string CellsEmptyMessage,
    IReadOnlyList<BatteryCellsStat> VoltageStats,
    bool ShowTemperature,
    IReadOnlyList<BatteryCellsStat> TemperatureStats)
{
    /// <summary>True when the heatmap has at least one cell (web <c>cells.length &gt; 0</c>).</summary>
    public bool HasCells => Cells.Count > 0;
}

/// <summary>
/// Pure projection from a raw <see cref="BatteryCellSummary"/> to the display model — the native port of
/// the <c>cellStatus</c> helper and the <c>statusCells</c>/stat <c>useMemo</c>s in
/// web/src/features/dashboard/widgets/BatteryCellsWidget.tsx. Voltages/temperatures are already SI and need
/// no conversion (the web widget shows raw volts / °), so this only formats and labels; every label resolves
/// through the i18n facade.
/// </summary>
public static class BatteryCellsProjection
{
    /// <summary>Fluent "Processor" glyph for the surface header / empty state (web <c>Cpu</c>).</summary>
    public const string HeaderGlyph = "\uE964";

    /// <summary>Deviation (mV) at or under which a cell is healthy (web <c>deviationMv &lt;= 5</c>).</summary>
    public const double OkThresholdMv = 5;

    /// <summary>Deviation (mV) at or under which a cell is a warning (web <c>deviationMv &lt;= 15</c>).</summary>
    public const double WarningThresholdMv = 15;

    private const string Degree = "\u00B0";

    /// <summary>
    /// Derive a cell's severity from how far its voltage deviates from the pack average (web
    /// <c>cellStatus</c>): missing voltage → <see cref="BatteryCellSeverity.Unknown"/>; otherwise
    /// ≤5&#160;mV → <see cref="BatteryCellSeverity.Ok"/>, ≤15&#160;mV → <see cref="BatteryCellSeverity.Warning"/>,
    /// else <see cref="BatteryCellSeverity.Error"/>.
    /// </summary>
    public static BatteryCellSeverity SeverityFor(double? voltage, double avgVoltage)
    {
        if (voltage is not { } v || double.IsNaN(v) || double.IsInfinity(v))
        {
            return BatteryCellSeverity.Unknown;
        }

        double deviationMv = Math.Abs(v - avgVoltage) * 1000;
        if (deviationMv <= OkThresholdMv)
        {
            return BatteryCellSeverity.Ok;
        }

        return deviationMv <= WarningThresholdMv ? BatteryCellSeverity.Warning : BatteryCellSeverity.Error;
    }

    /// <summary>Maps a cell severity to the shared semantic status used for its themed heatmap brush.</summary>
    public static StatusKind ToStatusKind(BatteryCellSeverity severity) => severity switch
    {
        BatteryCellSeverity.Ok => StatusKind.Success,
        BatteryCellSeverity.Warning => StatusKind.Warning,
        BatteryCellSeverity.Error => StatusKind.Danger,
        _ => StatusKind.Neutral,
    };

    /// <summary>Project <paramref name="data"/> for <paramref name="size"/> using the localizer for every label.</summary>
    public static BatteryCellsDisplay Project(
        BatteryCellSummary data,
        BatteryCellsSize size,
        ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(data);
        ArgumentNullException.ThrowIfNull(localizer);

        bool wide = size.IsWide;
        string cellWord = localizer.GetString("widget.batteryCells.cell", "Cell");
        double avgV = data.AvgVoltage;

        var cells = new List<BatteryCellStatusItem>(data.Cells.Count);
        foreach (var cell in data.Cells)
        {
            var severity = SeverityFor(cell.Voltage, avgV);
            string label = wide
                ? string.Create(CultureInfo.CurrentCulture, $"{cellWord} {cell.CellId} \u00B7 M{cell.Module}")
                : string.Create(CultureInfo.CurrentCulture, $"C{cell.CellId}");
            string value = wide
                ? $"{Fmt(cell.Voltage, 3)} V / {Fmt(cell.Temperature, 1)}{Degree}"
                : $"{Fmt(cell.Voltage, 3)} V";

            cells.Add(new BatteryCellStatusItem(
                Id: cell.CellId.ToString(CultureInfo.InvariantCulture),
                Label: label,
                Severity: severity,
                Value: value,
                AutomationName: $"{label}, {value}"));
        }

        var voltageStats = new List<BatteryCellsStat>(4)
        {
            new(localizer.GetString("widget.batteryCells.minV", "Min V"), $"{Fmt(data.MinVoltage, 3)} V"),
            new(localizer.GetString("widget.batteryCells.maxV", "Max V"), $"{Fmt(data.MaxVoltage, 3)} V"),
            new(localizer.GetString("widget.batteryCells.avgV", "Avg V"), $"{Fmt(avgV, 3)} V"),
            new(localizer.GetString("widget.batteryCells.spread", "Spread"), $"{Fmt(data.VoltageSpread * 1000, 1)} mV"),
        };

        IReadOnlyList<BatteryCellsStat> temperatureStats = wide
            ? new List<BatteryCellsStat>(3)
            {
                new(localizer.GetString("widget.batteryCells.minTemp", "Min Temp"), $"{Fmt(data.MinTemperature, 1)}{Degree}"),
                new(localizer.GetString("widget.batteryCells.avgTemp", "Avg Temp"), $"{Fmt(data.AvgTemperature, 1)}{Degree}"),
                new(localizer.GetString("widget.batteryCells.maxTemp", "Max Temp"), $"{Fmt(data.MaxTemperature, 1)}{Degree}"),
            }
            : Array.Empty<BatteryCellsStat>();

        return new BatteryCellsDisplay(
            IsCompact: size.IsCompact,
            IsWide: wide,
            GridColumns: size.GridColumns,
            Cells: cells,
            CellsEmptyMessage: localizer.GetString("widget.batteryCells.noCells", "No cell data"),
            VoltageStats: voltageStats,
            ShowTemperature: wide,
            TemperatureStats: temperatureStats);
    }

    /// <summary>
    /// Format a number exactly as the web <c>fmtNumber</c> does: coerce null / NaN / ±∞ to 0 (web
    /// <c>safeNumber</c>) then render with fixed <paramref name="decimals"/> fraction digits and en-US grouping.
    /// </summary>
    private static string Fmt(double? value, int decimals)
    {
        double safe = value is { } v && !double.IsNaN(v) && !double.IsInfinity(v) ? v : 0.0;
        return ScalarFormatters.FormatNumber(safe, decimals);
    }
}

/// <summary>
/// Maps the engine's raw <c>RepositoryResult&lt;JsonElement&gt;</c> emissions onto parsed
/// <c>RepositoryResult&lt;BatteryCellSummary&gt;</c>, preserving every freshness flag
/// (cached / refreshing / stale / offline) so the view-model can render the full state matrix. Kept pure so
/// the parse-and-preserve contract is unit-tested without a network or cache.
/// </summary>
public static class BatteryCellsResultMapper
{
    /// <summary>Parse <paramref name="raw"/>'s payload (when present) while preserving its status.</summary>
    public static RepositoryResult<BatteryCellSummary> Map(RepositoryResult<JsonElement> raw)
    {
        ArgumentNullException.ThrowIfNull(raw);

        BatteryCellSummary Parse() => raw.HasValue ? BatteryCellSummary.FromJson(raw.Value) : BatteryCellSummary.Empty;

        return raw.Status switch
        {
            LoadStatus.Loading => RepositoryResult<BatteryCellSummary>.Loading(),
            LoadStatus.Cached => RepositoryResult<BatteryCellSummary>.Cached(Parse(), raw.FetchedAt!.Value, raw.IsStale),
            LoadStatus.Refreshing => RepositoryResult<BatteryCellSummary>.Refreshing(Parse(), raw.FetchedAt!.Value, raw.IsStale),
            LoadStatus.Loaded => RepositoryResult<BatteryCellSummary>.Loaded(Parse(), raw.FetchedAt ?? DateTimeOffset.UtcNow),
            LoadStatus.Empty => RepositoryResult<BatteryCellSummary>.Empty(raw.FetchedAt),
            LoadStatus.Offline => RepositoryResult<BatteryCellSummary>.OfflineCached(Parse(), raw.FetchedAt!.Value, raw.Error!),
            _ => RepositoryResult<BatteryCellSummary>.Failure(
                raw.Error ?? new RepositoryError(RepositoryErrorKind.Unknown, "Unknown error")),
        };
    }
}

/// <summary>Null-tolerant numeric readers for the snake_case battery-cells JSON wire shape.</summary>
internal static class JsonNum
{
    /// <summary>Reads a numeric (or numeric-string) property, or null when absent / non-numeric.</summary>
    public static double? Double(JsonElement obj, string name)
    {
        if (!obj.TryGetProperty(name, out var v))
        {
            return null;
        }

        return v.ValueKind switch
        {
            JsonValueKind.Number when v.TryGetDouble(out var n) => n,
            JsonValueKind.String when double.TryParse(v.GetString(), NumberStyles.Float, CultureInfo.InvariantCulture, out var n) => n,
            _ => null,
        };
    }
}

File: apps/android/app/src/main/java/io/teslasync/android/dashboard/widgets/BatteryCellsWidgetModel.kt
package io.teslasync.android.dashboard.widgets

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull
import java.util.Locale
import kotlin.math.abs
import kotlin.math.roundToInt

/**
 * Per-cell severity the voltage heatmap colours a tile with — the native union of the web
 * `StatusCell['status']` values the shared `WidgetStatusGrid` renders. Derived from how far a
 * cell's voltage deviates from the pack average (web `cellStatus`): ≤5 mV → [Ok], ≤15 mV →
 * [Warning], >15 mV → [Error], missing voltage → [Unknown].
 */
enum class BatteryCellSeverity { Ok, Warning, Error, Unknown }

/**
 * One battery brick reading from `GET /vehicles/{vehicleID}/battery/cells` — the native mirror of
 * the web `BatteryCell` type (web/src/types/energy.ts): identity ([cellId], [module]) plus the
 * measured [voltage] (volts, SI) and [temperature] (°C, SI). Parsing is null-tolerant so a partial
 * body never throws; [voltage] stays `null` when the brick reported none so the heatmap renders the
 * web `'unknown'` status rather than a fabricated value.
 */
data class BatteryCell(
    val cellId: Int,
    val module: Int,
    val voltage: Double?,
    val temperature: Double?,
) {
    companion object {
        /** Project one cell object into a tolerant reading (binds to the web `BatteryCell` contract). */
        fun fromJson(element: JsonElement): BatteryCell {
            val obj = element as? JsonObject ?: return BatteryCell(0, 0, null, null)
            return BatteryCell(
                cellId = obj.numberOrNull("cell_id")?.roundToInt() ?: 0,
                module = obj.numberOrNull("module")?.roundToInt() ?: 0,
                voltage = obj.numberOrNull("voltage"),
                temperature = obj.numberOrNull("temperature"),
            )
        }
    }
}

/**
 * The battery-cell rollup from `GET /vehicles/{vehicleID}/battery/cells` (web `useBatteryCells`,
 * shape `BatteryCellSummary` in web/src/types/energy.ts). Field names mirror the Go API's snake_case
 * JSON tags. All voltages are volts and temperatures °C (SI); they are formatted for display only at
 * projection time and never unit-converted (the web widget shows raw volts / °, matching this).
 */
data class BatteryCellSummary(
    val avgVoltage: Double,
    val minVoltage: Double,
    val maxVoltage: Double,
    val voltageSpread: Double,
    val avgTemperature: Double,
    val minTemperature: Double,
    val maxTemperature: Double,
    val tempSpread: Double,
    val totalCells: Int,
    val cells: List<BatteryCell>,
) {
    companion object {
        /** An all-zero summary with no cells — the projection fallback for a populated-but-empty body. */
        val EMPTY: BatteryCellSummary =
            BatteryCellSummary(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0, emptyList())

        /**
         * Project a `GET /vehicles/{vehicleID}/battery/cells` body into a tolerant summary, or `null`
         * when the body is absent / not an object. Web parity: a present object (even all-zero) keeps
         * `data` truthy → the surface renders with the heatmap's own "No cell data" message; a missing
         * body → the outer "No battery cell data" empty state.
         */
        fun fromJson(element: JsonElement): BatteryCellSummary? {
            val obj = element as? JsonObject ?: return null
            return BatteryCellSummary(
                avgVoltage = obj.numberOrNull("avg_voltage") ?: 0.0,
                minVoltage = obj.numberOrNull("min_voltage") ?: 0.0,
                maxVoltage = obj.numberOrNull("max_voltage") ?: 0.0,
                voltageSpread = obj.numberOrNull("voltage_spread") ?: 0.0,
                avgTemperature = obj.numberOrNull("avg_temperature") ?: 0.0,
                minTemperature = obj.numberOrNull("min_temperature") ?: 0.0,
                maxTemperature = obj.numberOrNull("max_temperature") ?: 0.0,
                tempSpread = obj.numberOrNull("temp_spread") ?: 0.0,
                totalCells = obj.numberOrNull("total_cells")?.roundToInt() ?: 0,
                cells = (obj["cells"] as? JsonArray)?.map(BatteryCell::fromJson) ?: emptyList(),
            )
        }
    }
}

/**
 * The widget's grid footprint (columns × rows). Mirrors the web `WidgetProps.size` plus the
 * `isCompact` / `isWide` / heatmap-column logic in
 * web/src/features/dashboard/widgets/BatteryCellsWidget.tsx.
 */
data class BatteryCellsSize(
    val cols: Int,
    val rows: Int,
) {
    /** True at a single column (web `isCompact = size.cols <= 1`): hide the title, tighten the grid. */
    val isCompact: Boolean get() = cols <= COMPACT_MAX_COLS

    /** True at three or more columns (web `isWide = size.cols >= 3`): per-cell temperature + a temp row. */
    val isWide: Boolean get() = cols >= WIDE_MIN_COLS

    /** Heatmap column count (web `cols={isWide ? 4 : isCompact ? 2 : 3}`). */
    val gridColumns: Int
        get() =
            when {
                isWide -> WIDE_GRID_COLS
                isCompact -> COMPACT_GRID_COLS
                else -> STANDARD_GRID_COLS
            }

    companion object {
        private const val COMPACT_MAX_COLS = 1
        private const val WIDE_MIN_COLS = 3
        private const val COMPACT_GRID_COLS = 2
        private const val STANDARD_GRID_COLS = 3
        private const val WIDE_GRID_COLS = 4

        /** The registry default footprint (2×4). */
        val Default: BatteryCellsSize = BatteryCellsSize(2, 4)

        /** Minimum footprint (2×4) from the web registry. */
        val MinSize: BatteryCellsSize = BatteryCellsSize(2, 4)

        /** Maximum footprint (4×40) from the web registry. */
        val MaxSize: BatteryCellsSize = BatteryCellsSize(4, 40)

        /** True when [size] falls within the min/max footprint constraints. */
        fun withinBounds(size: BatteryCellsSize): Boolean =
            size.cols in MinSize.cols..MaxSize.cols && size.rows in MinSize.rows..MaxSize.rows

        /** Clamp [size] into the supported min/max footprint. */
        fun clamp(size: BatteryCellsSize): BatteryCellsSize =
            BatteryCellsSize(
                cols = size.cols.coerceIn(MinSize.cols, MaxSize.cols),
                rows = size.rows.coerceIn(MinSize.rows, MaxSize.rows),
            )
    }
}

/**
 * One projected, render-ready heatmap tile — the native analogue of a web `StatusCell`. Holds the
 * localized [label], the already-formatted [value], the derived [severity] (the view maps it to a
 * themed status colour) and a TalkBack [contentDescription]. Pure data — no Compose types.
 */
data class BatteryCellTile(
    val id: String,
    val label: String,
    val severity: BatteryCellSeverity,
    val value: String,
    val contentDescription: String,
)

/** One projected stat tile (min/max/avg/spread or per-module temperature) — web `StatCard`. */
data class BatteryCellsStat(
    val label: String,
    val value: String,
)

/**
 * The fully projected, render-ready view of the battery-cell summary for one footprint — the native
 * analogue of everything the web component computes via `useMemo` before returning JSX.
 */
data class BatteryCellsDisplay(
    val isCompact: Boolean,
    val isWide: Boolean,
    val gridColumns: Int,
    val cells: List<BatteryCellTile>,
    val voltageStats: List<BatteryCellsStat>,
    val showTemperature: Boolean,
    val temperatureStats: List<BatteryCellsStat>,
) {
    /** True when the heatmap has at least one cell (web `cells.length > 0`). */
    val hasCells: Boolean get() = cells.isNotEmpty()
}

/**
 * The localized labels the projection needs, resolved from the P1/S10 i18n catalog at the Compose
 * boundary (`stringResource`) and passed in so [BatteryCellsProjection.project] stays pure and
 * JVM-testable. Keys mirror the web `t('widget.batteryCells.*')` calls verbatim.
 */
data class BatteryCellsLabels(
    val cell: String,
    val minV: String,
    val maxV: String,
    val avgV: String,
    val spread: String,
    val minTemp: String,
    val avgTemp: String,
    val maxTemp: String,
)

/**
 * Pure projection from a raw [BatteryCellSummary] to the display model — the native port of the
 * `cellStatus` helper and the `statusCells`/stat `useMemo`s in
 * web/src/features/dashboard/widgets/BatteryCellsWidget.tsx. Voltages/temperatures are already SI
 * and need no conversion (the web widget shows raw volts / °), so this only formats and labels.
 */
object BatteryCellsProjection {
    /** Deviation (mV) at or under which a cell is healthy (web `deviationMv <= 5`). */
    const val OK_THRESHOLD_MV: Double = 5.0

    /** Deviation (mV) at or under which a cell is a warning (web `deviationMv <= 15`). */
    const val WARNING_THRESHOLD_MV: Double = 15.0

    private const val MV_PER_V = 1000.0
    private const val MIDDLE_DOT = "\u00B7"
    private const val DEGREE = "\u00B0"
    private const val VOLTAGE_DECIMALS = 3
    private const val TEMPERATURE_DECIMALS = 1
    private const val SPREAD_DECIMALS = 1

    /**
     * Derive a cell's severity from how far its voltage deviates from the pack average (web
     * `cellStatus`): missing/non-finite voltage → [BatteryCellSeverity.Unknown]; otherwise ≤5 mV →
     * [BatteryCellSeverity.Ok], ≤15 mV → [BatteryCellSeverity.Warning], else [BatteryCellSeverity.Error].
     */
    fun severityFor(
        voltage: Double?,
        avgVoltage: Double,
    ): BatteryCellSeverity {
        if (voltage == null || !voltage.isFinite()) return BatteryCellSeverity.Unknown
        val deviationMv = abs(voltage - avgVoltage) * MV_PER_V
        return when {
            deviationMv <= OK_THRESHOLD_MV -> BatteryCellSeverity.Ok
            deviationMv <= WARNING_THRESHOLD_MV -> BatteryCellSeverity.Warning
            else -> BatteryCellSeverity.Error
        }
    }

    /** Project [summary] for [size] using [labels] for every localized string. */
    fun project(
        summary: BatteryCellSummary,
        size: BatteryCellsSize,
        labels: BatteryCellsLabels,
    ): BatteryCellsDisplay {
        val wide = size.isWide
        val avgV = summary.avgVoltage
        val cells =
            summary.cells.map { cell ->
                val label =
                    if (wide) {
                        "${labels.cell} ${cell.cellId} $MIDDLE_DOT M${cell.module}"
                    } else {
                        "C${cell.cellId}"
                    }
                val value =
                    if (wide) {
                        "${fmt(cell.voltage, VOLTAGE_DECIMALS)} V / ${fmt(cell.temperature, TEMPERATURE_DECIMALS)}$DEGREE"
                    } else {
                        "${fmt(cell.voltage, VOLTAGE_DECIMALS)} V"
                    }
                BatteryCellTile(
                    id = cell.cellId.toString(),
                    label = label,
                    severity = severityFor(cell.voltage, avgV),
                    value = value,
                    contentDescription = "$label, $value",
                )
            }

        val voltageStats =
            listOf(
                BatteryCellsStat(labels.minV, "${fmt(summary.minVoltage, VOLTAGE_DECIMALS)} V"),
                BatteryCellsStat(labels.maxV, "${fmt(summary.maxVoltage, VOLTAGE_DECIMALS)} V"),
                BatteryCellsStat(labels.avgV, "${fmt(avgV, VOLTAGE_DECIMALS)} V"),
                BatteryCellsStat(labels.spread, "${fmt(summary.voltageSpread * MV_PER_V, SPREAD_DECIMALS)} mV"),
            )

        val temperatureStats =
            if (wide) {
                listOf(
                    BatteryCellsStat(labels.minTemp, "${fmt(summary.minTemperature, TEMPERATURE_DECIMALS)}$DEGREE"),
                    BatteryCellsStat(labels.avgTemp, "${fmt(summary.avgTemperature, TEMPERATURE_DECIMALS)}$DEGREE"),
                    BatteryCellsStat(labels.maxTemp, "${fmt(summary.maxTemperature, TEMPERATURE_DECIMALS)}$DEGREE"),
                )
            } else {
                emptyList()
            }

        return BatteryCellsDisplay(
            isCompact = size.isCompact,
            isWide = wide,
            gridColumns = size.gridColumns,
            cells = cells,
            voltageStats = voltageStats,
            showTemperature = wide,
            temperatureStats = temperatureStats,
        )
    }

    /**
     * Format a number exactly as the web `fmtNumber` does: coerce null / NaN / ±∞ to 0 (web
     * `safeNumber`) then render with fixed [decimals] fraction digits. Uses [Locale.US] to match the
     * web default locale and keep the projection deterministic.
     */
    fun fmt(
        value: Double?,
        decimals: Int,
    ): String {
        val safe = value?.takeIf(Double::isFinite) ?: 0.0
        return String.format(Locale.US, "%.${decimals}f", safe)
    }
}

/** Reads a numeric (or numeric-string) property, or `null` when absent / non-numeric. */
private fun JsonObject.numberOrNull(key: String): Double? = (this[key] as? JsonPrimitive)?.doubleOrNull
